### PR TITLE
feat(frontend): Dynamic GreedyBear news widget with filtering, sorting, and improved UX. Closes #762

### LINKED ISSUE #762: News feeds on start page are not generated dynamically
The GreedyBear news feeds displayed on the start page are currently hardcoded rather than being generated dynamically. It would be great if new content would appear automatically without code changes.

### PR DESCRIPTION
**Pull request by drona-gyawali**
# Description

This PR replaces the hardcoded news in Home.jsx with a dynamic NewsWidget component that fetches news from public/news.json. It filters, sorts, and displays only news related to GreedyBear, while preserving the original design.

Why news.json includes a project field and Id?
 - To ensure only GreedyBear news appears on the homepage.
 - Allows the JSON to include news for multiple projects in the future.
 - Filtering logic ensures future-proof dynamic loading without modifying code.
 -  React requires a unique key for each item when rendering a list.

## Related issues

closes:  #762 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

# Checklist

- [x] I have read and understood the rules about [how to Contribute](https://intelowlproject.github.io/docs/GreedyBear/Contribute/) to this project.
- [x] The pull request is for the branch `develop`.
- [ ] I have added documentation of the new features.
- [x] Linter (`Ruff`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowlproject.github.io/docs/GreedyBear/Contribute/#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [ ] I have added tests for the feature/bug I solved. All the tests (new and old ones) gave 0 errors.
- [ ] If changes were made to an existing model/serializer/view, the docs were updated and regenerated (check [CONTRIBUTE.md](https://github.com/intelowlproject/docs/blob/main/docs/GreedyBear/Contribute.md)).
- [x] If the GUI has been modified:
    - [x] I have a provided a screenshot of the result in the PR.
    - [ ] I have created new frontend tests for the new component or updated existing ones.
  
### Important Rules
- If you miss to compile the Checklist properly, your PR won't be reviewed by the maintainers.
- If your changes decrease the overall tests coverage (you will know after the Codecov CI job is done), you should add the required tests to fix the problem
- Everytime you make changes to the PR and you think the work is done, you should explicitly ask for a review. After being reviewed and received a "change request", you should explicitly ask for a review again once you have made the requested changes.

### CODE REVIEW

**Review by regulartim — CHANGES_REQUESTED** [review]
Hey @drona-gyawali ! :) Sorry, but I think I have not been precise enough when I wrote the issue. You created a local file that contains the links to the blog entries. But that's not what I actually want. 
All links are pointing to the IntelOwl blog https://intelowlproject.github.io/blogs . I would like to have a news widget that dynamically retrieves the links to GreedyBear related blog entries. Such that when a new article is published there, it should automatically appear on the GreedyBear start page. Do you understand what I mean?

If you don't feel like solving this issue anymore under that circumstances, that is fine, just tell me. 

(You linked the wrong issue in this PR. This PR should close #762 not #770 )

**Comment by drona-gyawali** [comment]
> Hey @drona-gyawali ! :) Sorry, but I think I have not been precise enough when I wrote the issue. You created a local file that contains the links to the blog entries. But that's not what I actually want. All links are pointing to the IntelOwl blog https://intelowlproject.github.io/blogs . I would like to have a news widget that dynamically retrieves the links to GreedyBear related blog entries. Such that when a new article is published there, it should automatically appear on the GreedyBear start page. Do you understand what I mean?
> 
> If you don't feel like solving this issue anymore under that circumstances, that is fine, just tell me.
> 
> (You linked the wrong issue in this PR. This PR should close #762 not #770 )


Extremely sorry for the confusion and for misunderstanding your request the first time.
I really appreciate your patience and the clear explanation you gave it helped me a lot.

To make sure I’m on the right track now:

What you want for the GreedyBear homepage news widget is:

 - Truly automatic / dynamic news: Whenever a new blog post appears on [IntelOwl blogs](https://intelowlproject.github.io/blogs)
 that is about GreedyBear, it should show up in the widget without anyone having to edit any file in the GreedyBear repo (no JSON file, no code change, nothing manual).

- The source of truth should be the live blog page itself (the one built from the .md files in the intelowlproject.github.io repo).

Only show posts clearly related to GreedyBear (like the ones with "GreedyBear" in the title: v3 coming, v2.0 released, improvements to GreedyBear, presenting GreedyBear, etc.).

I completely get why the news.json approach wasn’t ideal  it’s still semi-manual.

I understand the requirement and will push the new version soon.

**Comment by regulartim** [comment]
No worries, my issue describtion was too vague. Sorry for that. Yes, what you describe is exactly what I meant. If it is too hard to retrieve the list of blog posts from the IntelOwl website, you could also consider fetching them directly from the repo.

**Comment by drona-gyawali** [comment]
Hi @regulartim, I want to make sure I’m fully aligned before continuing implementation. So i did  some research on possible implementations and wanted to check alignment before coding further.

Thinking by  a long term perspective,  Frontend or backend HTML scraping/dom parsing feels brittle and hard to maintain also limitation of sorting by date.

A more reliable approach seems to be fetching GreedyBear-related blog posts directly from the IntelOwl GitHub repo via the GitHub API, parsing markdown with frontmatter on the backend, and exposing a small JSON endpoint for the frontend. This avoids dependency on the blog’s HTML structure and allows proper sorting/filtering and pagiantion and caching in future.


One small limitations, there’s a GitHub API rate limit for unauthenticated calls(only 60 req perhour), but this can be mitigated via Inmemory caching or an access token(5000 req per hour).

back-end response will look like this :
```
{
  "news": [
    {
      "title": "GreedyBear version 3 coming",
      "date": "2026-01-29",
      "link": "https://intelowlproject.github.io/blogs/greedybear_v3_release",
      "subtext": "Over the last months some new contributors helped us to implement a lot of new stuff in GreedyBear..."
    },

  ],
  "count": 1
}
```

what do you think ? does this approach sound acceptable to you, or would you prefer a different direction? And thank you very much for your valuable time on this pr.

**Comment by regulartim** [comment]
> Thinking by a long term perspective, Frontend or backend HTML scraping/dom parsing feels brittle and hard to maintain also limitation of sorting by date.

The sorting is not that important to me, but I agree, HTML scraping is not optimal. I was hoping that github.io has some kind of API where you can just retrieve a list of posts.

> One small limitations, there’s a GitHub API rate limit for unauthenticated calls(only 60 req perhour), but this can be mitigated via Inmemory caching or an access token(5000 req per hour).

Caching sounds good. If we cache for one hour, we would only have to make one request per hour per instance. Correct?

> what do you think ? does this approach sound acceptable to you, or would you prefer a different direction? And thank you very much for your valuable time on this pr.

As far as I can tell, this sounds very solid! :) Thank you for the time you spend investigating!

**Comment by drona-gyawali** [comment]
Hi,  In this version, I’ve reworked the implementation to align with the original intent of the issue, making the news widget fully automatic and truly dynamic.


The code in utils.py responsible for fetching news might look a bit unpolished because I prioritized keeping the logic within a single function. I’m happy to refactor it if you have suggestions on how you'd like it structured.


I have to  mention about  7-second latency that i notice in our api: 

 One GitHub API request

   - To list all files under Blogs/

Multiple sequential network requests

  - For each markdown file, we parse the raw .md content via download_url
  - Each request is a separate HTTPS round-trip to raw.githubusercontent.com

Network latency dominates  here Each request takes ~200–400 ms depending on network conditions With ~15–20 markdown files, this adds up quickly

Example:
```
1 list request        ~300 ms

15 markdown  ~15 × 300 ms ≈ 4.5–6 s

frontmatter parsing  ~negligible
```
Latency Benchmarks(approx):

    - Cold Start (Uncached): ~7s
    - Warm Start (Cached): 70ms – 162ms

However, Our  implementation has a room for scalability.  If our blog content grows or our user base increases, we can easily transition to a cron job model. This would handle the 'heavy lifting' (the 7s fetch) in the background every hour, ensuring that every user including the first one always receives a near-instant cached result.

Frontend 

Screenshot from 2026-02-06 22-37-38


Thank you for the review!

File: api/views/utils.py [hunk 1]
```diff
@@ -401,3 +405,75 @@ def asn_aggregated_queryset(iocs_qs, request, feed_params):
...
+    try:
+        # fetching the list of blog post files through github Api
+        response = requests.get(GITHUB_BLOG_API, timeout=10)
+        response.raise_for_status()
+        files = response.json()
+
+        news_items = []
+
+        # processing each markdown file to extract details
+        for file in files[:MAX_FILES_TO_CHECK]:
```
Comment (regulartim) [inline api/views/utils.py]: What do you think of filtering by filename pattern first here? Assume all GreedyBear related articles have the string "greedybear" in their name.

File: api/views/news.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+from rest_framework.decorators import api_view
+from rest_framework.response import Response
+
+from api.views.utils import get_greedybear_news
+
+
+@api_view(["GET"])
+def news_view(request):
```
Comment (regulartim) [inline api/views/news.py:8]: This API endpoint is for internal use only. Is there a decorator that can reflect this? Such that only our frontend can access the endpoint?

**Review by regulartim — CHANGES_REQUESTED** [review]
Hey @drona-gyawali ! Thanks a lot for your work and your assessment of the situation. I tested your code, and the widget looks very nice.

When I was looking at your code, another idea came to my mind which I want to discuss with you. We could also make some changes to the website https://intelowlproject.github.io . What if we would generate a RSS feed there and expose it as a `feed.xml` file to the internet. Then, instead of using the github API, we could just load the feed, which would contain all information that we need. Would this be the better approach to solve our issue here?

**Comment by drona-gyawali** [comment]
> When I was looking at your code, another idea came to my mind which I want to discuss with you. We could also make some changes to the website https://intelowlproject.github.io . What if we would generate a RSS feed there and expose it as a `feed.xml` file to the internet. Then, instead of using the github API, we could just load the feed, which would contain all information that we need. Would this be the better approach to solve our issue here?

Yes, using a feed.xml would be a great way to solve this. Having a single file to consume is much simpler and 10 times better than the current approach and would avoid the GitHub API entirely.

Even with this approach, caching would still be important yes, since fetching and parsing XML on every request can still be relatively expensive.

 I also saw that you’ve already opened a PR in another  repo once that gets merged, I’m happy to proceed with updating the implementation accordingly.

**Comment by regulartim** [comment]
Hi @drona-gyawali ! Good news, the PR got merged. :) Here is the link to the RSS feed: https://intelowlproject.github.io/feed.xml

File: api/views/news.py (hunk 2 above)
Comment (drona-gyawali) [reply api/views/news.py:8]: While doing some research, I found that Django or DRF as such  do not provide a built-in decorator to restrict an API to only our frontend. One possible approach is to send a token from the frontend and verify it on the backend; however, since this token is visible in developer tools, it cannot fully secure a public API.

For now, given that this is a public API, the most practical approach is to monitor its usage. If we observe frequent bot-like activity, we can implement rate limiting or block suspicious requests in the future.  And API isn’t documented, most users won’t find it, so casual exposure is limited.

A similar discussion on this topic can be found [here on Stack Overflow](https://stackoverflow.com/questions/47298148/how-to-ensure-that-only-my-frontend-can-access-my-api-and-not-anyone-else)
. Overall, while no solution is 100% secure for a public API . What do you think?

Comment (regulartim) [reply api/views/news.py:8]: Thanks for your investigation! :) In that case, just leave it as it is. Requests to this API should not be very resource intensive anyway, so I guess it is fine.

File: api/views/utils.py [hunk 3]
```diff
@@ -401,3 +405,63 @@ def asn_aggregated_queryset(iocs_qs, request, feed_params):
...
+                    "date": pub_date,
+                    "link": entry.get("link"),
+                    "subtext": subtext,
+                }
+            )
+
+        # RSS feeds are typically chronological, but we sort explicitly
+        # to ensure consistency since the data source is external.
+        news_items.sort(
+            key=lambda x: parsedate_to_datetime(x["date"]),
```
Comment (regulartim) [inline api/views/utils.py]: Feedparser already parses dates, every entry has a published_parsed attribute. I would prefer if you would use that instead. Then you can safely remove the (surprising) import from `email.utils`.

File: tests/test_views.py [hunk 4]
```diff
@@ -865,3 +870,152 @@ def test_is_sha256hash_invalid(self):
...
+        result = get_greedybear_news()
+        self.assertEqual(len(result), 2)
+        self.assertEqual(result[0]["title"], "GreedyBear New")
+        self.assertEqual(result[1]["title"], "GreedyBear Old")
+
+        # verifying sorting by date
+        self.assertGreater(
+            parsedate_to_datetime(result[0]["date"]),
+            parsedate_to_datetime(result[1]["date"]),
+        )
```
Comment (regulartim) [inline tests/test_views.py]: Same here maybe?

File: api/views/utils.py [hunk 5]
```diff
@@ -401,3 +405,63 @@ def asn_aggregated_queryset(iocs_qs, request, feed_params):
...
+    Returns:
+        List of dicts with keys: title, date, link, subtext
+        Sorted newest first, or empty list on failure.
+    """
+    cached = cache.get(CACHE_KEY_GREEDYBEAR_NEWS)
+    if cached is not None:
+        return cached
+
+    try:
+        feed = feedparser.parse(RSS_FEED_URL)
```
Comment (regulartim) [inline api/views/utils.py]: Here we are reading an external resource. What happens if it is not available for some reason? 

To my understanding, feedparser does not time out at any point. Maye it is safer to use `requests.get` with a timeout for fetching the contents and feedparser to parse the result?

Please also add a test for this.

Comment (drona-gyawali) [reply api/views/utils.py]: Yes, you are right feedparser uses urllib under the hood for http requests, which doesnot enforce timeout.  So I’ve updated the implementation to fetch the feed via requests.get with a timeout and pass the response content to feedparser. I also added a test that verifies we gracefully handle request timeouts by returning an empty list.

File: tests/test_views.py (hunk 4 above)
Comment (drona-gyawali) [reply tests/test_views.py]: I’ve updated the test to align with the new implementation. Since sorting now relies on published_parsed, the test simply verifies the order of results rather than re-parsing dates.

File: api/views/utils.py (hunk 3 above)
Comment (drona-gyawali) [reply api/views/utils.py]: I’ve updated the implementation to sort entries using published_parsed from feedparser and then build the API response using the human-readable published field. This removes the need for manual date parsing and keeps the response clean.

**Comment by drona-gyawali** [comment]
I noticed that test_views.py has already exceeded 1000 lines and is likely to grow further. To keep our tests maintainable and easier to navigate, I’m thinking of splitting it into separate test files (e.g., test_news.py). Should I open an issue and raise a PR for this in the next workflow?

**Comment by regulartim** [comment]
> Should I open an issue and raise a PR for this in the next workflow?

Yes, please! :)

**Review by regulartim — APPROVED** [review]
Good work! 👍

**Comment by drona-gyawali** [comment]
> Good work! 👍

Thanks! I’m really glad to hear that. I enjoyed working on this and I'm happy with how it turned out
